packages/@aws-cdk/core/lib/private/synthesis.ts

```diff
@@ -58,17 +58,25 @@ function synthNestedAssemblies(root: IConstruct, options: StageSynthesisOptions)
  * twice for the same construct.
  */
 function invokeAspects(root: IConstruct) {
+  let nestedAspectWarning = false;
   recurse(root, []);
 
   function recurse(construct: IConstruct, inheritedAspects: constructs.IAspect[]) {
     // hackery to be able to access some private members with strong types (yack!)
     const node: NodeWithAspectPrivatesHangingOut = construct.node._actualNode as any;
 
     const allAspectsHere = [...inheritedAspects ?? [], ...node._aspects];
-
+    const nodeAspectsCount = node._aspects.length;
     for (const aspect of allAspectsHere) {
       if (node.invokedAspects.includes(aspect)) { continue; }
       aspect.visit(construct);
+      // if an aspect was added to the node while invoking another aspect it will not be invoked, emit a warning
+      // the `nestedAspectWarning` flag is used to prevent the warning from being emitted for every child
+      if (!nestedAspectWarning && nodeAspectsCount !== node._aspects.length) {
+        // tslint:disable-next-line: max-line-length
+        construct.node.addWarning('We detected an aspect was added via another aspect, this is not supported and may result in an unexpected behavior');
```

Comment:
We can be more specific than this.

`Aspects added by other Aspects will not be applied; add all Aspects before 'synth' is called.`

Reply:
If we wanted to go with restoring the original behavior, it would look something like this:


```ts
const appliedHere = new Set<IAspect>();
let allAspectsHere: IAspect[];
while (true) {
  allAspectsHere = [...inheritedAspects ?? [], ...node._aspects];
  const yetToApply = allAspectsHere.filter(x => !appliedHere.has(x));
  if (yetToApply.length === 0) { break; }
  
  // ...
}

recurse(child, allAspectsHere);
```

Reply:
@eladb @rix0rrr I think we should restore the previous behavior, if only to maintain backward compatibility, if no objection I'll create the PR.

Reply:
No objections

Reply:
? Any updates?

Reply:
on it!

Reply:
Any updates?